hw/xbox/mcpx_apu.c

```diff
@@ -527,19 +581,24 @@ static void ep_write(void *opaque, hwaddr addr,
 {
     MCPXAPUState *d = opaque;
 
+    assert(size == 4);
+
     MCPX_DPRINTF("mcpx apu EP: [0x%llx] = 0x%llx\n", addr, val);
 
     switch (addr) {
-    case NV_PAPU_EPXMEM ... NV_PAPU_EPXMEM + 0xC00 * 4: {
-        assert(false);
+    case NV_PAPU_EPXMEM ... NV_PAPU_EPXMEM + 0xBFF*4: {
+        uint32_t xaddr = (addr - NV_PAPU_EPXMEM) / 4;
+        dsp_write_memory(d->ep.dsp, 'X', xaddr, val);
         break;
     }
-    case NV_PAPU_EPYMEM ... NV_PAPU_EPYMEM + 0x100 * 4: {
-        assert(false);
+    case NV_PAPU_EPYMEM ... NV_PAPU_EPYMEM + 0xFF*4: {
+        uint32_t yaddr = (addr - NV_PAPU_EPYMEM) / 4;
+        dsp_write_memory(d->ep.dsp, 'Y', yaddr, val);
         break;
     }
-    case NV_PAPU_EPPMEM ... NV_PAPU_EPPMEM + 0x1000 * 4: {
-        assert(false);
+    case NV_PAPU_EPPMEM ... NV_PAPU_EPPMEM + 0xFFF*4: {
```

Comment:
These offsets strike me as odd... These implicit switch cases are inclusive right? Should it not be `NV_PAPU_EPPMEM ... NV_PAPU_EPPMEM + 0x1000*4-1`

Reply:
Well.. it's kind-of the same. As we assert 32 bit access it makes no difference (actually the `- 1` solution could write more data than expected then).

I just modified existing code, but I also knew we do this same sort of stuff in the GPU:

```c
case NV097_SET_VERTEX4F ... NV097_SET_VERTEX4F + 12: {
```

*(where 4 values = 16 bytes, can be passed to NV097_SET_VERTEX4F. So the formula is `(4-1)*4`)*

I can do the `<size> * 4 - 1` if you want that.

Please tell me if you want this changed. I personally think we should introduce macros at some point in the future and do an overall cleanup.

Reply:
In this case it may not cause an issue because the factor is 4, but for larger values this certainly would cause a problem. Consider if instead we had used a factor of 8, then we would lose the ability to address the last dword.

For correctness and readability, I'd prefer it (for now) if we did:

```c
case <base> ... <base> + <size> * 4 - 1:
```

But yes, I do agree that at some point we should probably cleanup this switch stuff and use a macro or some table lookup thing.